File: conformance/reference-python/query_serialization.py
import math
from decimal import Decimal


UNRESERVED = frozenset(b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~")


def failure(code):
    return {"ok": False, "code": code}


def is_valid_unicode_scalar_string(value):
    if not isinstance(value, str):
        return False
    return all(not 0xD800 <= ord(character) <= 0xDFFF for character in value)


def percent_encode(value):
    output = []
    for byte in value.encode("utf-8"):
        output.append(chr(byte) if byte in UNRESERVED else f"%{byte:02X}")
    return "".join(output)


def decode_base_query_component(value):
    output = bytearray()
    index = 0
    try:
        while index < len(value):
            if value[index] == "%":
                if index + 2 >= len(value):
                    return None
                output.append(int(value[index + 1:index + 3], 16))
                index += 3
            else:
                output.extend(value[index].encode("utf-8"))
                index += 1
        decoded = output.decode("utf-8", errors="strict")
        return decoded if is_valid_unicode_scalar_string(decoded) else None
    except (UnicodeError, ValueError):
        return None


def jcs_number(value):
    number = float(value)
    if not math.isfinite(number):
        return None
    if number == 0:
        return "0"

    shortest = repr(number).lower()
    absolute = abs(number)
    if 1e-6 <= absolute < 1e21:
        fixed = format(Decimal(shortest), "f")
        if "." in fixed:
            fixed = fixed.rstrip("0").rstrip(".")
        return fixed

    mantissa, exponent = shortest.split("e")
    if mantissa.endswith(".0"):
        mantissa = mantissa[:-2]
    exponent_value = int(exponent)
    exponent_text = f"+{exponent_value}" if exponent_value >= 0 else str(exponent_value)
    return f"{mantissa}e{exponent_text}"


def scalar_to_text(value):
    if value is None:
        return {"tombstone": True}
    if isinstance(value, str):
        return {"text": value} if is_valid_unicode_scalar_string(value) else None
    if isinstance(value, bool):
        return {"text": "true" if value else "false"}
    if isinstance(value, (int, float)):
        text = jcs_number(value)
        return {"text": text} if text is not None else None
    return None


def serialize_query(base_url, sources):
    if not isinstance(base_url, str) or not isinstance(sources, list):
        return failure("INVALID_QUERY_INPUT")

    fragment_index = base_url.find("#")
    request_part = base_url if fragment_index == -1 else base_url[:fragment_index]
    fragment = "" if fragment_index == -1 else base_url[fragment_index:]
    query_index = request_part.find("?")
    path = request_part if query_index == -1 else request_part[:query_index]
    base_query = "" if query_index == -1 else request_part[query_index + 1:]
    merged = {}

    for segment in base_query.split("&"):
        if segment == "":
            continue
        encoded_key, separator, encoded_value = segment.partition("=")
        if separator == "":
            encoded_value = ""
        key = decode_base_query_component(encoded_key)
        value = decode_base_query_component(encoded_value)
        if key is None or value is None:
            return failure("INVALID_BASE_URL_QUERY")
        if key == "":
            return failure("INVALID_QUERY_KEY")
        merged[key] = value

    for source in sources:
        if not isinstance(source, list):
            return failure("INVALID_QUERY_INPUT")
        for entry in source:
            if not isinstance(entry, list) or len(entry) != 2:
                return failure("INVALID_QUERY_INPUT")
            key, value = entry
            if not is_valid_unicode_scalar_string(key) or key == "":
                return failure("INVALID_QUERY_KEY")
            scalar = scalar_to_text(value)
            if scalar is None:
                return failure("INVALID_QUERY_VALUE")
            if scalar.get("tombstone"):
                merged.pop(key, None)
            else:
                merged[key] = scalar["text"]

    query = "&".join(
        f"{percent_encode(key)}={percent_encode(merged[key])}"
        for key in sorted(merged, key=lambda item: tuple(ord(character) for character in item))
    )
    return {"ok": True, "url": f"{path}{'' if query == '' else '?' + query}{fragment}"}

File: conformance/reference-js/query-serialization.js
'use strict';

function failure(code) {
  return { ok: false, code };
}

function isValidUnicodeScalarString(value) {
  for (let index = 0; index < value.length; index += 1) {
    const codeUnit = value.charCodeAt(index);
    if (codeUnit >= 0xd800 && codeUnit <= 0xdbff) {
      const next = value.charCodeAt(index + 1);
      if (!(next >= 0xdc00 && next <= 0xdfff)) return false;
      index += 1;
    } else if (codeUnit >= 0xdc00 && codeUnit <= 0xdfff) {
      return false;
    }
  }
  return true;
}

function compareUnicodeCodePoints(left, right) {
  const leftPoints = Array.from(left, character => character.codePointAt(0));
  const rightPoints = Array.from(right, character => character.codePointAt(0));
  const length = Math.min(leftPoints.length, rightPoints.length);
  for (let index = 0; index < length; index += 1) {
    if (leftPoints[index] !== rightPoints[index]) return leftPoints[index] - rightPoints[index];
  }
  return leftPoints.length - rightPoints.length;
}

function percentEncode(value) {
  return encodeURIComponent(value).replace(/[!'()*]/g, character =>
    `%${character.charCodeAt(0).toString(16).toUpperCase()}`,
  );
}

function decodeBaseQueryComponent(value) {
  try {
    const decoded = decodeURIComponent(value);
    return isValidUnicodeScalarString(decoded) ? decoded : null;
  } catch {
    return null;
  }
}

function scalarToText(value) {
  if (value === null || value === undefined) return { tombstone: true };
  if (typeof value === 'string') {
    return isValidUnicodeScalarString(value) ? { text: value } : null;
  }
  if (typeof value === 'boolean') return { text: value ? 'true' : 'false' };
  if (typeof value === 'number' && Number.isFinite(value)) {
    return { text: Object.is(value, -0) ? '0' : JSON.stringify(value) };
  }
  return null;
}

function serializeQuery(baseUrl, sources) {
  if (typeof baseUrl !== 'string' || !Array.isArray(sources)) {
    return failure('INVALID_QUERY_INPUT');
  }

  const fragmentIndex = baseUrl.indexOf('#');
  const requestPart = fragmentIndex === -1 ? baseUrl : baseUrl.slice(0, fragmentIndex);
  const fragment = fragmentIndex === -1 ? '' : baseUrl.slice(fragmentIndex);
  const queryIndex = requestPart.indexOf('?');
  const path = queryIndex === -1 ? requestPart : requestPart.slice(0, queryIndex);
  const baseQuery = queryIndex === -1 ? '' : requestPart.slice(queryIndex + 1);
  const merged = new Map();

  for (const segment of baseQuery.split('&')) {
    if (segment === '') continue;
    const equalsIndex = segment.indexOf('=');
    const encodedKey = equalsIndex === -1 ? segment : segment.slice(0, equalsIndex);
    const encodedValue = equalsIndex === -1 ? '' : segment.slice(equalsIndex + 1);
    const key = decodeBaseQueryComponent(encodedKey);
    const value = decodeBaseQueryComponent(encodedValue);
    if (key === null || value === null) return failure('INVALID_BASE_URL_QUERY');
    if (key.length === 0) return failure('INVALID_QUERY_KEY');
    merged.set(key, value);
  }

  for (const source of sources) {
    if (!Array.isArray(source)) return failure('INVALID_QUERY_INPUT');
    for (const entry of source) {
      if (!Array.isArray(entry) || entry.length !== 2) return failure('INVALID_QUERY_INPUT');
      const [key, value] = entry;
      if (typeof key !== 'string' || key.length === 0 || !isValidUnicodeScalarString(key)) {
        return failure('INVALID_QUERY_KEY');
      }
      const scalar = scalarToText(value);
      if (scalar === null) return failure('INVALID_QUERY_VALUE');
      if (scalar.tombstone) merged.delete(key);
      else merged.set(key, scalar.text);
    }
  }

  const query = [...merged.entries()]
    .sort(([left], [right]) => compareUnicodeCodePoints(left, right))
    .map(([key, value]) => `${percentEncode(key)}=${percentEncode(value)}`)
    .join('&');

  return { ok: true, url: `${path}${query === '' ? '' : `?${query}`}${fragment}` };
}

module.exports = { serializeQuery };